### DISCUSSION TITLE
Email templates

### DISCUSSION THREAD

**Discussion by Beetix** [ory/kratos#906] [Ideas]
Hi,

I've made my custom email templates and I was able to receive emails generated from them.

For now they are very basic containing just a bit of text. I'm now looking at making something a bit more elaborate using HTML.

I would like to create a common layout for all emails. Looking at the current code it does seem possible to share template code between files.

Diving into Go nested templates gave me an idea to make it more flexible. Why not add a "common" directory and add all the files it contains to the parsing with the function [`ParseFiles`](https://golang.org/pkg/text/template/#ParseFiles)? The user could then reuse common blocks for the various mails.

For example:
`common/layout.gotmpl`
```
{{define "base"}}
<html>
<body>
{{template "content" .}}
</body>
</html>
{{end}}
```

`recovery/valid/email.body.gotmpl`
```
{{define "content"}}
Recovery
{{end}}

{{template "base"}}
```

`verification/valid/email.body.gotmpl`
```
{{define "content"}}
Verification
{{end}}

{{template "base"}}
```
...
What do you think?

**Comment by aeneasr** [comment]
I think that's definitely an option - however we want to keep customising templates as straight forward as possible. The fewer layout tricks we use (such as includes) the easier it is to change things. I think one common problem with these type of emails is that HTML is very difficult to get right in email, as the email clients all handle this stuff differently. One interesting approach would be to support both HTML and text templates.

By the way, I was thinking that it makes sense to have an email template gallery of sorts? What do you think about that?

**Reply by Beetix** [reply]
Yes, what I had in mind would allow both simple plain text and more complex HTML mails with code reuse.

An email template gallery seems also a good idea.

**Comment by emanmks** [comment]
Please also consider to give us the flexibility to use the templates from our 3rd party email service provider (mailgun, SES, etc). Some teams preferring this approach (to manage the templates from the email service and not customize in the code).
IMHO, this means we are going to support a non-SMTP interface with the email service provider.

**Reply by aeneasr** [reply]
Yes, this was always an idea we had. The problem is a bit that there are many of such providers and we don't want to maintain the solutions for everyone of those. Especially because developers have many ideas around customization which leads to a lot of config options usually. Is there a standard we could use instead?

**Comment by zepatrik** [comment]
Please also have a look at https://github.com/ory/kratos/issues/834 as it is somehow related. I think loading a whole directory that you can use and reuse templates from would be nice. But low complexity is definitely the first priority.

**Comment by jay-kanakala** [comment]
Hey Beetix, how did you do that? custom email templates thing. In which folder did you keep the templates and how did you give the path in kratos,yml file. Any kindof tutorial or github repo would be very helpful. documentation says some /conf/courier-templates/ folder.... but where is that folder? should I create it? how did you do it man, pls tell me
